courses/models.py

```diff
@@ -0,0 +1,119 @@
+from django.db import models
+from utils.constants import SKILL_BEGINNER, SKILL_INTERMEDIATE, SKILL_ADVANCED, EXTERNAL_LINK_BLOG, EXTERNAL_LINK_TUTORIAL, EXTERNAL_LINK_COURSE
+from django.utils.text import slugify
+# Create your models here.
+
+class Language(models.Model):
+    language_name = models.CharField(max_length=20)
+    slug = models.SlugField(unique=True)
+
+    def _get_unique_slug(self):
+        slug = slugify(self.language_name)
+        unique_slug = slug
+        num = 1
+        while Language.objects.filter(slug=unique_slug).exists():
+            unique_slug = '{}-{}'.format(slug, num)
+            num += 1
+        return unique_slug
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = self._get_unique_slug()
+        super(Language, self).save(*args, **kwargs)
+
+    def __str__(self):
+        return '{name}-{slug}'.format(name=self.language_name, slug=self.slug)
+
+
+class Domain(models.Model):
+    domain_name = models.CharField(max_length=20)
+    slug = models.SlugField(unique=True)
+
+    def _get_unique_slug(self):
+        slug = slugify(self.domain_name)
+        unique_slug = slug
+        num = 1
+        while Domain.objects.filter(slug=unique_slug).exists():
+            unique_slug = '{}-{}'.format(slug, num)
+            num += 1
+        return unique_slug
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = self._get_unique_slug()
+        super(Domain, self).save(*args, **kwargs)
+
+    def __str__(self):
+        return '{name}-{slug}'.format(name=self.domain_name, slug=self.slug)
+
+class RowInformation(models.Model):
+    is_active = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    modified_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        abstract = True
+
+
+class KnowledgeBase(RowInformation):
+    skill_levels = (
+        (SKILL_BEGINNER, 'BEGINNER'),
+        (SKILL_INTERMEDIATE, 'INTERMEDIATE'),
+        (SKILL_ADVANCED, 'ADVANCED')
+    )
+
+    title = models.CharField(max_length=50, null=False, blank=False)
+    description = models.TextField(null=False, blank=True)    #if no description, store empty string
+    slug = models.SlugField(unique=True)
+    languages = models.ManyToManyField(Language, related_name='knowledgebase_lang')
+    domains = models.ManyToManyField(Domain, related_name='knowledgebase_dom')
+    skill_level = models.CharField(max_length=2, choices=skill_levels)
+
+    class Meta:
+        abstract = True
+        ordering = ['-modified_at','-created_at']
+
+    def _get_unique_slug(self):
+        slug = slugify(self.title)
+        unique_slug = slug
+        num = 1
+        while KnowledgeBase.objects.filter(slug=unique_slug).exists():
```

Comment:
```py
import this
"""
The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
"""
```
Even though my code looks complex, your code seems complicated (*to me*)

In your code, we are doing one query every time we are saving in DB. The will be huge a red flag